### PR TITLE
AIX support for System.Native/PAL

### PR DESCRIPTION
**Pull request by NattyNarwhal**
Somewhat incomplete, but at least on Mono version, gets it compiling,
so work on it can be done; as I believe the nopped out functions aren't
hooked up to Mono's BCL yet. For now, if merged, would let me get back
to working on a patch for UnicodeEncoding ;)

See mono/corefx#69 for the description.

This is for Mono, as CoreCLR doesn't run under AIX or I. (Mono downstream would prefer if this is merged here.)

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @janvorli

File: src/Native/Unix/System.Native/pal_errno.c [hunk 1]
```diff
@@ -130,8 +130,10 @@ int32_t SystemNative_ConvertErrorPlatformToPal(int32_t platformErrno)
             return Error_ENOTCONN;
         case ENOTDIR:
             return Error_ENOTDIR;
+#if !defined(_AIX)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_errno.c]: Can this instead be:
```C++
#ifdef ENOTEMPTY
```
so that it's not specific to _AIX and rather just applies wherever the symbol is defined?

File: src/Native/Unix/System.Native/pal_io.c [hunk 2]
```diff
@@ -78,6 +90,8 @@ c_static_assert(PAL_S_IFSOCK == S_IFSOCK);
 
 // Validate that our enum for inode types is the same as what is
 // declared by the dirent.h header on the local system.
+// (AIX doesn't have dirent d_type, so none of this there)
+#if !defined(_AIX)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.c]: Could this be done with:
```C++
#ifdef DT_UNKNOWN
```
or some such thing?

File: src/Native/Unix/System.Native/pal_io.c [hunk 3]
```diff
@@ -233,14 +248,20 @@ static int32_t ConvertOpenFlags(int32_t flags)
             return -1;
     }
 
+#if defined(_AIX)
+    if (flags & ~(PAL_O_ACCESS_MODE_MASK | PAL_O_CREAT | PAL_O_EXCL | PAL_O_TRUNC | PAL_O_SYNC))
+#else
     if (flags & ~(PAL_O_ACCESS_MODE_MASK | PAL_O_CLOEXEC | PAL_O_CREAT | PAL_O_EXCL | PAL_O_TRUNC | PAL_O_SYNC))
+#endif
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.c]: Why do we need this ifdef?  PAL_O_CLOEXEC is defined regardless of the native dependencies.

File: src/Native/Unix/System.Native/pal_io.c [hunk 4]
```diff
@@ -233,14 +248,20 @@ static int32_t ConvertOpenFlags(int32_t flags)
...
     {
         assert_msg(false, "Unknown Open flag", (int)flags);
         return -1;
     }
 
+#if !defined(_AIX)
     if (flags & PAL_O_CLOEXEC)
         ret |= O_CLOEXEC;
+#endif
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.c]: As in other places, could this instead be done with something like:
```C++
#ifdef O_CLOEXEC
```
?

File: src/Native/Unix/System.Native/pal_io.c [hunk 5]
```diff
@@ -265,6 +286,14 @@ intptr_t SystemNative_Open(const char* path, int32_t flags, int32_t mode)
 
     int result;
     while ((result = open(path, flags, (mode_t)mode)) < 0 && errno == EINTR);
+#if defined(_AIX)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.c]: Same.  I'll stop commenting on these.

File: src/Native/Unix/System.Native/pal_networking.c [hunk 6]
```diff
@@ -39,15 +41,19 @@
 #endif
 #include <unistd.h>
 #include <pwd.h>
+#if !defined(_AIX)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: Why is this necessary?  If _AIX doesn't have these, shouldn't the cmake configuration just be setting HAVE_SENDFILE_4/6 to 0?

File: src/Native/Unix/System.Native/pal_networking.c [hunk 7]
```diff
@@ -39,15 +41,19 @@
...
 #if HAVE_SENDFILE_4
 #include <sys/sendfile.h>
 #elif HAVE_SENDFILE_6
 #include <sys/uio.h>
 #endif
+#endif
 #if !HAVE_IN_PKTINFO
 #include <net/if.h>
+#if !defined(_AIX)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: Can we do this instead with cmake configuration?

File: src/Native/Unix/System.Native/pal_networking.c [hunk 8]
```diff
@@ -782,6 +788,8 @@ static int32_t GetIPv4PacketInformation(struct cmsghdr* controlMessage, struct I
     ConvertInAddrToByteArray(&packetInfo->Address.Address[0], NUM_BYTES_IN_IPV4_ADDRESS, &pktinfo->ipi_addr);
 #if HAVE_IN_PKTINFO
     packetInfo->InterfaceIndex = (int32_t)pktinfo->ipi_ifindex;
+#elif defined(_AIX)
+    // TODO: SIOCGIFCONF based fallback
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: I'd prefer it if we avoid adding these TODOs.  Since this is incomplete anyway, let's just leave them out and the appropriate fallbacks can be added when implementation is available for them.

File: src/Native/Unix/System.Native/pal_networking.c [hunk 9]
```diff
@@ -2413,7 +2446,7 @@ int32_t SystemNative_SendFile(intptr_t out_fd, intptr_t in_fd, int64_t offset, i
     *sent = 0;
     return SystemNative_ConvertErrorPlatformToPal(errno);
 
-#elif HAVE_SENDFILE_6
+#elif HAVE_SENDFILE_6 && !defined(_AIX) /* why is AIX getting this? */
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: The fix should be to update the cmake configuration however is necessary, not to add this !defined(_AIX) here.

File: src/Native/Unix/System.Native/pal_random.c [hunk 10]
```diff
@@ -40,7 +40,12 @@ void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int3
 
             do
             {
+#ifdef O_CLOEXEC
                 fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+#else
+                fd = open("/dev/urandom", O_RDONLY, NULL);
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_random.c]: Why not:
```C++
open("/dev/urandom", O_RDONLY);
```
?

File: src/Native/Unix/System.Native/pal_errno.c (hunk 1 above)
Comment (janvorli) [reply src/Native/Unix/System.Native/pal_errno.c]: +1, this should be detected at build time in the configure.cmake and get HAVE_ENOTEMPTY defined to 0 or 1.

File: src/Native/Unix/System.Native/pal_io.c [hunk 11]
```diff
@@ -325,7 +354,30 @@ static void ConvertDirent(const struct dirent* entry, struct DirectoryEntry* out
     // the start of the unmanaged string. Give the caller back a pointer to the
     // location of the start of the string that exists in their own byte buffer.
     outputEntry->Name = entry->d_name;
+#if defined(_AIX)
+    /* AIX has no d_type, make a substitute */
+    struct stat s;
+    stat(entry->d_name, &s);
+    if (S_ISDIR(s.st_mode)) {
```
Comment (janvorli) [inline src/Native/Unix/System.Native/pal_io.c]: A nit - the coding convention is to place the opening curly bracket at a new line.

File: src/Native/Unix/System.Native/pal_random.c [hunk 12]
```diff
@@ -40,7 +40,12 @@ void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int3
 
             do
             {
+#ifdef O_CLOEXEC
```
Comment (janvorli) [inline src/Native/Unix/System.Native/pal_random.c]: I would prefer detecting this in the configure.cmake too. The O_CLOEXEC doesn't have to be a define and in such case, this check would break.

File: src/Native/Unix/System.Native/pal_errno.c (hunk 1 above)
Comment (NattyNarwhal) [reply src/Native/Unix/System.Native/pal_errno.c]: AIX aliases ENOTEMPTY to EEXIST:

```c
#if defined(_ALL_SOURCE) && !defined(_LINUX_SOURCE_COMPAT)
#define ENOTEMPTY       EEXIST  /* Directory not empty */
#else   /* not _ALL_SOURCE */
#define ENOTEMPTY       87
#endif  /* _ALL_SOURCE */
```

File: src/Native/Unix/System.Native/pal_io.h [hunk 13]
```diff
@@ -15,6 +15,17 @@ BEGIN_EXTERN_C
 #include <dirent.h>
 #include <sys/types.h>
 
+#if defined(_AIX) && !defined(O_CLOEXEC)
+// HACK: Try to get AIX 6.1 & i 7.1 working; AIX 7.1 & i (claims) 7.2 has these but we target
+// these older versions too.
+// We ifdef it out for the open call in .c and simulate with fcntl, but for SystemNative_Pipe,
+// since we also don't have pipe2 on AIX/i, it'll fall back to using FD_CLOEXEC.
+// (i 7.2 at least defines O_CLOEXEC as 0x0000001000000000LL)
+#define 0x0000001000000000LL
```
Comment (janvorli) [inline src/Native/Unix/System.Native/pal_io.h]: This define is missing the symbol.

File: src/Native/Unix/System.Native/pal_errno.c (hunk 1 above)
Comment (stephentoub) [reply src/Native/Unix/System.Native/pal_errno.c]: Then could it be:
```C++
if ENOTEMPTY != EEXIST
```
?

**Comment by NattyNarwhal** [comment]
WRT the sendfile call: I'm attached to the Mono build machinery, not CoreRT cmake, and for some reason it's thinking sendfile is valid?

good point on ifaddrs; I know the Mono machinery detects that case; I'm just not sure how to add it to CMake.

**Comment by stephentoub** [comment]
> I'm attached to the Mono build machinery, not CoreRT cmake, and for some reason it's thinking sendfile is valid?

We should understand and fix that on the Mono side then rather than complicating the source.

**Comment by NattyNarwhal** [comment]
```
configure:26246: checking for sendfile with 6 arguments
configure:26266: gcc -c -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -g -O2 -std=gnu99 -fno-strict-aliasing -fwrapv -DMONO_DLL_EXPORT -Wno-unused-but-set-variable -g -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -D__PASE__ -I/opt/mono/include -maix64 -DGC_AIX_THREADS -D_ALL_SOURCE -D_THREAD_SAFE -D_LARGE_FILES -D_REENTRANT -g  conftest.c >&5
conftest.c:251:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 main ()
 ^~~~
conftest.c: In function 'main':
conftest.c:254:16: warning: implicit declaration of function 'sendfile' [-Wimplicit-function-declaration]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
                ^~~~~~~~
conftest.c:254:3: warning: nested extern declaration of 'sendfile' [-Wnested-externs]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
   ^~~
conftest.c:254:7: warning: unused variable 'result' [-Wunused-variable]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
       ^~~~~~
configure:26266: $? = 0
configure:26268: result: yes
```

I'll file this part at the Mono end. Will fix up the ifdefs now.

**Comment by NattyNarwhal** [comment]
WRT to `O_CLOEXEC`; is there an implementation that defines it as a constant instead of a preprocessor definition?

I'm also unsure as to where to place the case for the ifaddrs check if that's not present already in the CMakeLists, even though I'm not building with those; I know it is on Mono.

Meanwhile, figuring out the Mono sendfile-in-autoconf issue.

**Comment by janvorli** [comment]
> WRT to O_CLOEXEC; is there an implementation that defines it as a constant instead of a preprocessor definition?

I am not aware of any, however, it is preferable to not to assume things that are not documented. 
On the other hand, I can see that corefx mini-PALs already have couple of places where similar checks are done, so I guess I am ok with it staying as is.

File: src/Native/Unix/System.Native/pal_io.h [hunk 14]
```diff
@@ -15,6 +15,12 @@ BEGIN_EXTERN_C
 #include <dirent.h>
 #include <sys/types.h>
 
+#if defined(_AIX) && !defined(F_DUPFD_CLOEXEC)
+// HACK: Try to get AIX 6.1 & i 7.1 working; AIX 7.1 & i (claims) 7.2 has these but we target
+// these older versions too, so nop this define out.
+#define F_DUPFD_CLOEXEC 0
```
Comment (janvorli) [inline src/Native/Unix/System.Native/pal_io.h]: I am not sure how this is going to work. We use this symbol at one place only:
https://github.com/dotnet/corefx/blob/09b3cc02d85c5c963af50567cb89b6d67beebc49/src/Native/Unix/System.Native/pal_io.c#L276-L281

I am not sure how defining it as 0 would work there.

Comment (NattyNarwhal) [reply src/Native/Unix/System.Native/pal_io.h]: Yeah, you're right; I wrote that when I was really ill and irritated. Let me fix that.

File: src/Native/Unix/System.Native/pal_io.h [hunk 15]
```diff
@@ -267,12 +267,22 @@ enum SysConfName
...
 {
+#if defined(_AIX)
+/* of course AIX is different */
+    PAL_POLLIN = 0x0001,   /* non-urgent readable data available */
+    PAL_POLLPRI = 0x0004,  /* urgent readable data available */
+    PAL_POLLOUT = 0x0002,  /* data can be written without blocked */
+    PAL_POLLERR = 0x4000,  /* an error occurred */
+    PAL_POLLHUP = 0x2000,  /* the file descriptor hung up */
+    PAL_POLLNVAL = 0x8000, /* the requested events were invalid */
+#else
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.h]: This it not correct.  These values are the PAL values and must be identical to the ones used in the managed definition:
https://github.com/dotnet/corefx/blob/d9d06b0109c600cd7a13a8fa5783d8e6b0c745a4/src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs#L14-L23

You will need to create the appropriate mapping when these are used.

Comment (NattyNarwhal) [reply src/Native/Unix/System.Native/pal_io.h]: Would the mapping need to be implemented on the C# side (figure out a way to ifdef on it) or translate the system values to the PAL ones et vice versa?

Comment (stephentoub) [reply src/Native/Unix/System.Native/pal_io.h]: No, on the native side.  The idea is that the managed code always uses a fixed set of values (the platform abstraction layer "PAL" values) when communicating with the native shim.  The native shim is then compiled on each platform where there are differences, and it's responsible for ensuring that the managed code is always able to use those PAL values, so if the PAL values aren't the same such that it can't just pass them through, the native code needs to do the translation.

File: src/Native/Unix/System.Native/pal_io.c [hunk 16]
```diff
@@ -37,6 +42,13 @@
 #include <sys/inotify.h>
 #endif
 
+#if defined(_AIX)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.c]: Above you used `#ifdef _AIX`, here you used `#if defined(_AIX)`.  It'd be good to be consistent.  These files are already predominantly using `#ifdef` and `#ifndef`... can we use that for all of these new checks?

File: src/Native/Unix/System.Native/pal_networking.c [hunk 17]
```diff
@@ -2247,7 +2251,32 @@ static int32_t WaitForSocketEventsInner(int32_t port, struct SocketEvent* buffer
 }
 
 #else
-#error Asynchronous sockets require epoll or kqueue support.
+// TODO: Poll fallback
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: Is this something you plan to follow-up on?  If no, please remove the TODO.  If yes, please open an issue for it and include the issue number here.

File: src/Native/Unix/System.Native/pal_random.c (hunk 12 above)
Comment (stephentoub) [reply src/Native/Unix/System.Native/pal_random.c]: The nice thing about doing this in configure.cmake is that we can easily see in the build output / CI logs what choice it's making.  If this ifdef were to fail for some unexpected reason, the code would fall back to the worse path, and we wouldn't know it.

File: src/Native/Unix/System.Native/pal_networking.c (hunk 17 above)
Comment (NattyNarwhal) [reply src/Native/Unix/System.Native/pal_networking.c]: I plan to, but since it involves quite a bit of work and I don't think anything on the Mono side uses this yet, not in this PR.

I'll remove this and make the defined stuff more consistent.

File: src/Native/Unix/configure.cmake [hunk 18]
```diff
@@ -87,6 +87,20 @@ check_c_source_compiles(
     "
     HAVE_FLOCK64)
 
+check_symbol_exists(
```
Comment (janvorli) [inline src/Native/Unix/configure.cmake]: You also need to add the new symbols to https://github.com/dotnet/corefx/blob/master/src/Native/Unix/Common/pal_config.h.in as `#cmakedefine01`

**Comment by janvorli** [comment]
@NattyNarwhal one more issue:
```
07:22:27 /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/Native/Unix/System.Native/pal_io.c:125:9: error: macro name must be an identifier
07:22:27 #ifndef (_AIX)
07:22:27         ^
07:22:27 1 error generated.
```
The parentheses should not be there.

**Comment by stephentoub** [comment]
@janvorli, any other comments?

File: src/Native/Unix/System.Native/pal_networking.c [hunk 19]
```diff
@@ -20,6 +20,8 @@
 #include <sys/types.h>
 #include <sys/event.h>
 #include <sys/time.h>
+#else
+#include <sys/poll.h>
```
Comment (janvorli) [inline src/Native/Unix/System.Native/pal_networking.c]: Is the sys/poll.h guaranteed to exist on systems that don't have epoll or kqueue? If not, then I would also recommend adding a detection for the header presence to configure.cmake and include the header only if it exists.

File: src/Native/Unix/configure.cmake [hunk 20]
```diff
@@ -316,6 +316,10 @@ check_function_exists(
     fcopyfile
     HAVE_FCOPYFILE)
 
+check_function_exists(
```
Comment (janvorli) [inline src/Native/Unix/configure.cmake]: I rather meant checking for the sys/poll.h header presence to make sure we don't try to include it if it doesn't exist. Like this:
```cmake
check_include_files(
     "sys/poll.h"
     HAVE_SYS_POLL_H)
```

I am not sure if the poll function detection as you have it would work, since I think that cmake would not add the "sys/poll.h" include into its little testing source that it generates.
